test/unit-tests/ima_test.js

```diff
@@ -119,20 +187,703 @@ describe('ad_manager_ima', function() {
+      //current placeholder id is undefined, update this when this changes
+      if(typeof adId === "undefined") {
+        notified = true;
+      }
+    };
+    initialize(true);
+    //This line mimics the ad pod having been set by the time the ad request success returns
+    //This is a hack, will revisit in the future, but does properly test to see that the fake ad
+    //ends when an ad request is successful
+    ima.currentAMCAdPod = amc.timeline[0];
+    play();
+    expect(notified).to.be(true);
+  });
+
+  // Non-Ad Rules
+  it('Init, Non-Ad Rules: ad is added to timeline for non-ad rules ads', function(){
     initialize(false);
     expect(amc.timeline.length).to.be(1);
     expect(amc.timeline[0].ad.type).not.to.be("adRequest");
   });
 
-  it('Init: fake ad is added to timeline for ad rules ads', function(){
-    debugger;
+  it('Init, Non-Ad Rules: setup ads request is successful', function(){
+    initialize(false);
+    play();
+    ima.playAd(amc.timeline[0]);
+    expect(ima.adsRequested).to.be(true);
+  });
+
+  // AMC integration/IMA Event tests
+  it('AMC Integration, Non-Ad Rules: amc is notified of a non-linear ad playback', function(){
+    var notified = false;
+    google.ima.linearAds = false;
+    initAndPlay(false, vci);
+    var id = "blah";
+    var adPod = {
+      id : id,
+      ad : {
+        forced_ad_type : amc.ADTYPE.NONLINEAR_OVERLAY
+      }
+    };
+    amc.sendURLToLoadAndPlayNonLinearAd = function(currentAdPod, adPodId, url) {
+      if (adPod === currentAdPod && id === adPodId) {
+        notified = true;
+      }
+    };
+    ima.playAd(adPod);
+    expect(notified).to.be(true);
+  });
+
+  it('AMC Integration, IMA Event: IMA CLICK event notifies amc of an ad click', function(){
+    var notified = false;
+    initAndPlay(true, vci);
+    amc.adsClicked = function() {
+      notified = true;
+    };
+    ima.playAd({
+      ad : {}
+    });
+    var am = google.ima.adManagerInstance;
+    am.publishEvent(google.ima.AdEvent.Type.CLICK);
+    expect(notified).to.be(true);
+  });
+
+  it('AMC Integration, IMA Event: IMA AD_ERROR gives back control and ends current ad and ad pod', function(){
+    var doneControllingAdsNotified = false;
+    var linearAdEndedNotified = false;
+    var podEndedNotified = false;
+    initAndPlay(true, vci);
+    amc.notifyLinearAdEnded = function() {
+      linearAdEndedNotified = true;
+    };
+
+    amc.notifyPodEnded = function() {
+      podEndedNotified = true;
+    };
+
+    amc.adManagerDoneControllingAds = function(adManagerName) {
+      if (adManagerName === name) {
+        doneControllingAdsNotified = true;
+      }
+    };
+    ima.playAd({
+      ad : {}
+    });
+    var am = google.ima.adManagerInstance;
+    am.publishEvent(google.ima.AdEvent.Type.STARTED);
+    am.publishEvent(google.ima.AdEvent.Type.AD_ERROR);
+    expect(linearAdEndedNotified).to.be(true);
+    expect(podEndedNotified).to.be(true);
+    expect(doneControllingAdsNotified).to.be(true);
```

Comment:
I think we need to clean up functions that you add to the amc for a specific test. Or does the amc get recreated for each test?